ankidoc.1
.TH ANKIDOC 1 2024-12-17 "ankidoc"
.
.SH NAME

ankidoc \- convert asciidoc notes to anki notes
.
.SH SYNOPSIS
.
.SY ankidoc
.B \-h, \-\-help
.YS
.
.SY ankidoc
.RB "[ " \-d " | " \-l " | " \-n " ]"
.RB [ \-o
.IR OUT ]
.RB [ \-a
.IR ATTR ]
.RB [ \-L
.IR LV ]
.I files
.
.SH DESCRIPTION
.
\fBankidoc\fP is a short python script intended to make writing anki notes in
asciidoc easier. It can be used to generate a text file for import into anki
from notes written in asciidoc.
.
.P
The asciidoc notes need to be in a specific format in order for the script to
be able to work with them; The idea is that each note \(em or at least each of
the notes the user wants to have in one deck \(em is assigned a unique ID. It
is then split into two asciidoc files, one with the content of the front side
of the note, and one of the back side. These two files must now be saved as
\fIID.front\fP and \fIID.back\fP, ID being the unique note ID that was assigned
to the note.
.
.P
Then the list of \fI.front\fP files desired to be in the final import file can
be passed to \fBankidoc\fP. The script will then find the corresponding
\fI.back\fP files (which have to be in the same directory), convert all files
to HTML using the
.BR asciidoctor (1)
utility and output a properly formatted anki import file named \fIout\fP. The
name and location of the output file can be adjusted using the \fB-o\fP option.
.
.SS MODES
.
Besides the default mode described above, \fBankidoc\fP supports additional
modes that work slightly differently.
.
.TP
.B Notegen mode
In notegen mode \(em selected by specifying the \fB-n\fP option \(em
\fBankidoc\fP will only take in one single front file, find its corresponding
back file, convert both sides and output a \(dqnote\(dq file, which essentially
corresponds to a single note how it would appear in the final import file
generated in default mode. These note files can then be linked into an anki
import file using the link mode.
.
.TP
.B Link mode
In link mode \(em selected by specifying the \fB-l\fP option \(em \fBankidoc\fP
will treat the input files as \(dqnote\(dq files and link them into an anki
import file.
.
.TP
.B Docgen mode
In docgen mode \(em selected by specifying the \fB-d\fP option \(em
\fBankidoc\fP will, instead of generating a import file, generate a HTML
document out of the notes. This mode can, for example, be used to preview the
notes before importing them into anki.
.
.SH OPTIONS
.
.TP
.BI "\-a, \-\-attributes " ATTR
Specify any asciidoc(tor) attributes. Whatever string is specified here will be
passed on as\-is to the \fB-a\fP argument of the 
.BR asciidoctor (1)
command.
.
.TP
.B \-d, \-\-docgen
Run \fBankidoc\fP in docgen mode.
.
.TP
.B \-l, \-\-link
Run \fBankidoc\fP in link mode.
.
.TP
.BI "\-L, \-\-loglevel " LV
Select a logging level from \(dqDEBUG\(dq, \(dqINFO\(dq, \(dqWARNING\(dq,
\(dqERROR\(dq, \(dqCRITICAL\(dq.
.
.TP
.B \-n, \-\-notegen
Run \fBankidoc\fP in notegen mode.
.
.TP
.BI "\-o, \-\-output " OUT
Specify the desired output filename.
.
.TP
.I files
Specify the files to operate on. Note that these generally have to be
\fI.front\fP files with existing corresponding \fI.back\fP files in oder for
\fBankidoc\fP to work. This is except when operating in link mode, in which
case this option has to be one single \fI.note\fP file.
.
.SH EXIT STATUS
.
.TP
.B 0
Success.
.
.TP
.B 1
Failure.
.
.SH NOTES
For \fBankidoc\fP to work you'll have to have
.BR asciidoctor (1)
installed, as this is the program \fBankidoc\fP uses to convert the asciidoc to
HTML.
.
.SH SEE ALSO
.
.nh
.BR asciidoctor (1)
.
.P
.UR https://asciidoctor.org
.UE


ankidoc.py
#!/usr/bin/env python3

import argparse
import logging
import subprocess
import os
import sys

anki_header = """#separator:semicolon
#html:true
#columns:id;question;answer

"""

# Construct a asciidoctor command.
def get_adoc_cmd(embedded, attributes, input, output):
    cmd = ["asciidoctor"]

    if embedded:
        cmd.append("-e")

    if attributes:
        cmd.append("-a")
        cmd.append(attributes)

    cmd.append("-o")
    cmd.append(output)
    cmd.append(input)

    logging.debug(f"constructed command \"{cmd}\"");

    return cmd

# Pass a stderr output of a subprocess to the logging system.
def pass_stderr(stderr):

    if stderr == None or stderr == b'':
        return

    stderr_string = stderr.decode("utf-8")
    stderr_lines = stderr_string.splitlines()

    for line in stderr_lines:
        logging.warning(line)

# Generate a note from a front file.
def notegen(front_path, attributes):
    logging.info(f"running notegen on {front_path}")

    id_path, ext = os.path.splitext(front_path)

    if ext != ".front":
        logging.warning(f"{front_path} is not a front file, skipping")
        return None

    id = os.path.basename(id_path)
    back_path = id_path + ".back"

    front = subprocess.run(get_adoc_cmd(True, attributes, front_path, "-"), capture_output=True)
    pass_stderr(front.stderr)

    back = subprocess.run(get_adoc_cmd(True, attributes, back_path, "-"), capture_output=True)
    pass_stderr(back.stderr)

    if front.stdout == None or front.stdout == b'' or back.stdout == None or back.stdout == b'':
        return None

    front_contents = front.stdout.decode("utf-8").replace("\"", "\"\"")
    back_contents = back.stdout.decode("utf-8").replace("\"", "\"\"")

    return f"\"{id}\";\"{front_contents}\";\"{back_contents}\"\n"

# Link notes into one output file.
def link(note_paths, output_path):
    logging.info("running linker")
    logging.debug(f"notes to link: {note_paths}")

    with open(output_path, "w") as output:
        output.write(anki_header)

    with open(output_path, "a") as output:

        for note_path in note_paths:

            if not os.path.exists(note_path):
                logging.warning(f"{note_path} doesn't exist, not linked")
                continue
            elif not os.path.splitext(note_path)[1] == ".note":
                logging.warning(f"{note_path} not a note file, not linked")
                continue

            logging.info(f"linking {note_path}")

            with open(note_path, "r") as note_file:
                output.write(note_file.read())

# Run the program in default mode on 'front_paths'.
def default_mode(front_paths, output_path, attributes):

    logging.debug(f"operating on {front_paths}")

    with open(output_path, "w") as output:
        output.write(anki_header)

    with open(output_path, "a") as output:

        for front_path in front_paths:

            note = notegen(front_path, attributes)

            if not note == None:
                output.write(note)

# Run the program in docgen mode on 'front_paths'.
def docgen_mode(front_paths, output_path, attributes):

    logging.debug(f"operating on {front_paths}")

    asciidoc_output = ""

    for front_path in front_paths:

        id_path, ext = os.path.splitext(front_path)

        if ext != ".front":
            logging.warning(f"{front_path} is not a front file, skipping")
            continue

        id = os.path.basename(id_path)
        back_path = id_path + ".back"

        if not os.path.isfile(front_path):
            logging.warning(f"{front_path} is not a file, skipping")
            continue
        elif not os.path.isfile(back_path):
            logging.warning(f"{back_path} is not a file, skipping")
            continue

        front_contents = ""
        back_contents = ""

        with open(front_path, "r") as front_file:
            front_contents = front_file.read()

        with open(back_path, "r") as back_file:
            back_contents = back_file.read()

        asciidoc_output += f"\n{front_contents}\n\n_{id}_\n\n{back_contents}\n"

    asciidoc_bytes = asciidoc_output.encode("utf-8")

    asciidoctor = subprocess.run(get_adoc_cmd(False, attributes, "-", output_path), input=asciidoc_bytes, capture_output=True)
    pass_stderr(asciidoctor.stderr)

# Run the program in notegen mode on 'front_paths'.
def notegen_mode(front_paths, output_path, attributes):

    logging.debug(f"operating on {front_paths}")

    if len(front_paths) > 1:
        logging.critical("cannot run in notegen mode when generating multiple files")
        exit(1)

    note = notegen(front_paths[0], attributes)

    if note == None:
        return

    with open(output_path, "w") as note_file:
        note_file.write(note)

# Run the program in link mode on 'front_paths'.
def link_mode(front_paths, output_path):

    logging.debug(f"operating on {front_paths}")

    link(front_paths, output_path)

def main():

    # Initialise the argument parser and all arguments.
    parser = argparse.ArgumentParser(
        description="convert asciidoc notes to anki notes",
        usage="%(prog)s [-h] [ -d | -n | -l ] [-o OUT] [-a ATTR] [-L LV] files ..."
    )

    parser.add_argument(
        "-a", "--attributes",
        metavar="ATTR",
        help="any asciidoctor attributes to set"
    )

    parser.add_argument(
        "-d", "--docgen",
        action="store_true",
        help="generate a single html document from the front files passed"
    )

    parser.add_argument(
        "-l", "--link",
        action="store_true",
        help="link the note files passed into one anki import file"
    )

    parser.add_argument(
        "-L", "--loglevel",
        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
        default="WARNING",
        metavar="LV",
        help="select a logging level"
    )

    parser.add_argument(
        "-n", "--notegen",
        action="store_true",
        help="compile the front/back files passed into note files"
    )

    parser.add_argument(
        "-o", "--output",
        default="out",
        metavar="OUT",
        help="the desired output filename"
    )

    parser.add_argument(
        "files",
        nargs="*",
        help="the files to operate on"
    )

    args = parser.parse_args()

    # Configure the logging mechanism.
    loglevel = logging.WARNING

    if args.loglevel == "DEBUG":
        loglevel = logging.DEBUG
    elif args.loglevel == "INFO":
        loglevel = logging.INFO
    elif args.loglevel == "WARNING":
        loglevel = logging.WARNING
    elif args.loglevel == "ERROR":
        loglevel = logging.ERROR
    elif args.loglevel == "CRITICAL":
        loglevel = logging.CRITICAL

    logging.basicConfig(format=f"{parser.prog}: %(levelname)s: %(message)s", level=loglevel)

    # Rule out common errors.
    if (args.docgen and args.notegen) or (args.docgen and args.link) or (args.notegen and args.link):
        logging.critical("incompatible modes")
        exit(1)

    if args.files == []:
        logging.critical("no files provided")
        exit(1)

    # Run the program in the mode requested by the user.
    if args.docgen:
        docgen_mode(args.files, args.output, args.attributes)
    elif args.notegen:
        notegen_mode(args.files, args.output, args.attributes)
    elif args.link:
        link_mode(args.files, args.output)
    else:
        default_mode(args.files, args.output, args.attributes)

    exit(0)

if __name__ == "__main__":
    main()
